### PR TITLE
Generalized support points / interpolation

### PR DESCRIPTION
**Pull request by tamiko**
Work in progress - do not merge.

In reference to #4704 (aka the last pull request ;-) )

In reference to #4992

### CODE REVIEW

**Comment by tamiko** [comment]
@luca-heltai This is what I have so far - your input please! :-)

**Comment by tamiko** [comment]
@luca-heltai I haven't found time for some more serious testing - but so far this method returns (up to round-off error)  the same result as `VectorTools::interpolate`.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 1]
```diff
@@ -85,6 +85,208 @@ DEAL_II_NAMESPACE_OPEN
...
+    // initialize a quadrature with all generalized support points and
+    // create an FEValues object with it.
+
+    hp::QCollection<dim> support_quadrature;
+    for (unsigned int fe_index=0; fe_index<fe.size(); ++fe_index)
+      {
+        const auto &points = fe[fe_index].get_generalized_support_points();
+        const auto size = points.size();
+        std::vector<double> weights(size, 1. / size);
+        support_quadrature.push_back(Quadrature<dim>(points, weights));
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: since you don't need the weights, just create the quadrature with only the points -- there is a constructor for that.

Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: @bangerth I need valid JxW values (even though the weight doesn't matter). Unfortunately, said constructor initializes the weights with infinity. This took me 2h to figure out yesterday.

Comment (bangerth) [reply include/deal.II/numerics/vector_tools.templates.h]: Hm. What do you need the JxW values for? If the weight doesn't matter, isn't all you need the determinant of the Jacobian?

Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: Good point. I have updated the code.

**Comment by tamiko** [comment]
@luca-heltai @bangerth Please have another look - this is pretty much the function I want to merge (at some point).

There is quite a bit of maintenance still that has to be done.

I want to integrate above logic into `VectorTools::interpolate` that either uses the slightly simpler old code for Lagrangian interpolation, or above variant if necessary.

File: include/deal.II/fe/fe.h [hunk 2]
```diff
@@ -1951,13 +1951,15 @@ class FiniteElement : public Subscriptor,
    */
   bool has_generalized_support_points () const;
 
+//FIXME
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/fe/fe.h]: Agreed.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 3]
```diff
@@ -85,6 +85,263 @@ DEAL_II_NAMESPACE_OPEN
...
+        dof_values.resize(n_dofs);
+
+        // Get all function values:
+        function.vector_value_list(generalized_support_points,
+                                   function_values);
+
+        // Before we can average, we have to transform all function values
+        // from the real cell back to the unit cell. We query the finite
+        // element for the correct transformation. This is not really ideal
+        // but the best we can currently do...
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: let me rephrase this comment.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 4]
```diff
@@ -85,6 +85,263 @@ DEAL_II_NAMESPACE_OPEN
...
+    //    according to the conformity of the component (scalar, Hdiv, or
+    //    Hcurl conforming); see [Monk, Finite Element Methods for Maxwell's
+    //    Equations, p.77ff Section 3.9] for details. This results in
+    //    \hat u(\hat x_i)
+    //
+    //  - We convert these generalized support point values to nodal values
+    //
+    //  - For every global dof we take the average 1 / n_K \sum_{K} dof_K
+    //    where n_K is the number of cells sharing the global dof and dof_K
+    //    is the computed value on the cell K.
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: After a short skype discussion I think we sorted out that point :-)

**Comment by tamiko** [comment]
Currently failing tests: http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=11242

**Comment by tamiko** [comment]
@bangerth @luca-heltai I think this PR is in a good state now.

I want to add one final test for the Nedelec element (but this can also happen in a separate PR).

There are a couple of improvement left that I want to address in separate PRs over the course of the next weeks. (Such as support for pulling back mixed vector spaces, etc.).

File: include/deal.II/numerics/vector_tools.templates.h [hunk 5]
```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
...
+              fe[fe_index].get_nonzero_components(i);
 
-              // get indices of the dofs on this cell
-              dofs_on_cell.resize (fe[fe_index].dofs_per_cell);
-              cell->get_dof_indices (dofs_on_cell);
+            // Figure out whether the component mask applies. We assume
+            // that we are allowed to set degrees of freedom if at least
+            // one of the components (of the dof) is selected.
+            //
+            // FIXME: Add a debug assert.
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: add one, maybe?

**Comment by bangerth** [comment]
I haven't really tried to fully understand what the algorithm does -- I assume you and Luca already figured this out. I just pointed out some style issues.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 6]
```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
...
+    // Now loop over all locally owned, active cells.
+    //
 
-    for (unsigned int fe_index=0; fe_index<fe.size(); ++fe_index)
+    for (auto cell : dof_handler.active_cell_iterators())
       {
-        for (unsigned int i=0; i<fe[fe_index].dofs_per_cell; ++i)
+        // If this cell is not locally owned, do nothing.
+        if (! cell->is_locally_owned())
+          continue;
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply include/deal.II/numerics/vector_tools.templates.h]: Not really. The other processor will add to the global vector just as we do. This follows the same principle of assembly. We don't assemble on ghost cells (because our neighbor will)...

File: include/deal.II/fe/fe_tools.templates.h [hunk 7]
```diff
@@ -2916,6 +2916,79 @@ namespace FETools
...
+        double_support_point_values.get().resize(support_point_values.size());
+        double_dof_values.get().resize(dof_values.size());
+
+        for (unsigned int i = 0; i < support_point_values.size(); ++i)
+          {
+            double_support_point_values.get()[i].reinit(
+              finite_element.n_components(), false);
+            std::copy(std::begin(support_point_values[i]),
+                      std::end(support_point_values[i]),
+                      std::begin(double_support_point_values.get()[i]));
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply include/deal.II/fe/fe_tools.templates.h]: for example `std::complex<double>`, or sacado type objects, or AdolC, etc.

[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/fe/fe_tools.templates.h]: But isn't the additional information lost in that case? Also, a `std::complex` can not be copied to a `double`. So it's not clear to me how this would make it possible to interpolate complex-valued initial conditions.

File: include/deal.II/numerics/vector_tools.templates.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/numerics/vector_tools.templates.h]: Ah, yes, of course. Doh.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 8]
```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
...
+            //  \hat p = p\circ F_K
+            //  i.e., do nothing.
+            //
+            break;
 
-    std::vector<types::global_dof_index> dofs_on_cell (fe.max_dofs_per_cell());
+          default:
+            // In case we deal with an unknown conformity, just assume we
+            // deal with a Lagrange element and do nothing.
+            break;
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: In one of the next iterations of this function, I want to support this type of FESystems.

File: include/deal.II/fe/fe_tools.h [hunk 9]
```diff
@@ -557,13 +557,31 @@ namespace FETools
...
+   * If the number type is double this function simply calls the
+   * FiniteElement::convert_generalized_support_point_values_to_dof_values()
+   * directly. Otherwise, temporary (per-thread storage) is used to call to
+   * above function.
+   */
+  template <int dim, int spacedim, typename number>
+  void convert_generalized_support_point_values_to_dof_values(
+    const FiniteElement<dim, spacedim> &finite_element,
+    const std::vector<Vector<number> > &support_point_values,
+    std::vector<number>                &nodal_values);
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/fe/fe_tools.h]: done

File: include/deal.II/fe/fe_tools.templates.h [hunk 10]
```diff
@@ -2916,6 +2916,79 @@ namespace FETools
...
 
 
+  namespace
+  {
+    // internal helper class to partially specialize the number type
+
+    template <int dim, int spacedim, typename number>
+    struct ConvertHelper
+    {
+      static void convert_generalized_support_point_values_to_dof_values(
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/fe/fe_tools.templates.h]: done

File: include/deal.II/fe/fe_tools.h [hunk 11]
```diff
@@ -557,13 +557,31 @@ namespace FETools
...
+
+
+  /**
+   * Wrapper around
+   * FiniteElement::convert_generalized_support_point_values_to_dof_values()
+   * that works with arbitrary number types.
+   *
+   * If the number type is double this function simply calls the
+   * FiniteElement::convert_generalized_support_point_values_to_dof_values()
+   * directly. Otherwise, temporary (per-thread storage) is used to call to
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/fe/fe_tools.h]: done

File: include/deal.II/fe/fe_tools.templates.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/fe/fe_tools.templates.h]: @bangerth The variant for `std::complex<double>` is simply not yet implemented.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 12]
```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
...
+        if (fe[fe_index].dofs_per_cell == 0)
+          continue;
+
+        // Get transformed, generalized support points
+        fe_values.reinit(cell);
+        const std::vector<Point<spacedim> > &generalized_support_points =
+          fe_values.get_present_fe_values().get_quadrature_points();
+
+        // Get indices of the dofs on this cell
+        const auto n_dofs = fe[fe_index].dofs_per_cell;
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: Well, I strongly disagree. Actually, I do not know what the return type is. Is it `unsigned int`, is it `size_t`? Did we plan for the future and use `types::global_dof_index`?

The point is it doesn't really matter.

File: include/deal.II/numerics/vector_tools.templates.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: (also @luca-heltai) For this to work out we will have to implement a `Mapping::inverse_transform` first. Further, we will have to make the mapping type of the (base) elements available in the finite element.

The current solution is a bit of a hack around that problem.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 13]
```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
...
-    hp::QCollection<dim> support_quadrature;
-    for (unsigned int fe_index=0; fe_index<fe.size(); ++fe_index)
-      support_quadrature.push_back (Quadrature<dim>(rep_unit_support_points[fe_index]));
+          } /*switch*/
 
-    // Transformed support points are computed by
-    // FEValues
-    hp::MappingCollection<dim,spacedim> mapping_collection (mapping);
+        FETools::convert_generalized_support_point_values_to_dof_values(
+          fe[fe_index], function_values, dof_values);
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: added a note.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 14]
```diff
@@ -85,215 +85,284 @@ DEAL_II_NAMESPACE_OPEN
...
 
-    hp::FEValues<dim,spacedim> fe_values (mapping_collection,
-                                          fe, support_quadrature, update_quadrature_points);
+        for (unsigned int i=0; i < n_dofs; ++i)
+          {
+            ::dealii::internal::ElementAccess<VectorType>::add(
+              typename VectorType::value_type(1.0),
+              dofs_on_cell[i],
+              weights);
```
[the remark this answers is not in the document]
Comment (tamiko) [reply include/deal.II/numerics/vector_tools.templates.h]: But doesn't the same hold true for writing directly into `vec` and `interpolant`?

I.e., I don't see how the old version of this function (that directly inserted into `vec`) would have worked in that case either.

[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/numerics/vector_tools.templates.h]: Fair point. The previous version would also have failed.

**Comment by tamiko** [comment]
@bangerth 
I have thought about the issue of the local conversion function and threat safety - ultimately it boils down to that we really want `FiniteElement::convert_generalized_support_point_values_to_dof_values` to never schedule tbb tasks. For starters it is conceptually wrong - it is a relatively lightweight operation and parallelizing on sub-cell data is too fine-grained.

But more importantly, I really want to use a lock-free, per-thread storage. The reason is that I plan to use `WorkStream` to parallelize the interpolation in a follow-up pull-request and intermediate storage for this call is in a "hot path" (so a global mutex for a vector pool is costly).

Yes, I can restructure the code in a lock-free, reentrant version with per-thread storage. But I would like to avoid that (parallelizing the per cell task as explained above will not gain any speed increase).

The small vector object doesn't really work because we store a `dealii::Vector` inside that uses heap memory.

**Comment by tamiko** [comment]
This pull request currently breaks two remaining tests:

 - `bits/step-15`, solver has different initial/final residuals due to interpolation of starting values

 - `dofs/interpolate_q_system_mixed_01`, the FESystem in this test contains a non-interpolatory element, consequently the system does not have any generalized support points. I am not sure how to recover the old behavior of ignoring a non-interpolatory base element if the component mask doesn't select it.

**Comment by tamiko** [comment]
Apart from that, I would be very happy if we could consider merging the current state and incrementally improving / fixing the rest. (The pull request is already way to large.)

**Comment by luca-heltai** [comment]
I agree. Let's just merge this one and then work our way through next PRs.

**Comment by bangerth** [comment]
> I have thought about the issue of the local conversion function and threat 
> safety - ultimately it boils down to that we really want 
> |FiniteElement::convert_generalized_support_point_values_to_dof_values| to 
> never schedule tbb tasks. For starters it is conceptually wrong - it is a 
> relatively lightweight operation and parallelizing on sub-cell data is too 
> fine-grained.

I know that's what you want, but you can't control that because you're calling 
functions for which you don't know what they do. For example, `vec = 0` 
creates threads. My point was that your code is not thread safe if you call 
*anything*.


> But more importantly, I really want to use a lock-free, per-thread storage. 

The data structure we use in WorkStream is lock free. You just need to have a 
(per-thread) pool of objects from which every thread can choose one that is 
not currently in use.

**Comment by tamiko** [comment]
@bangerth Do you see any chance for this to be merged without fixing the threading issue if `convert_generalized_support_point_values_to_dof_values` happens to use `dealii::Vector`?

I have already piled up another 3 pull requests on top of this one.

**Comment by bangerth** [comment]
Yes, let's merge if you open a github issue that reminds us of the issue.

/run-tests

**Comment by bangerth** [comment]
Btw, I don't think it makes a difference what vector type you use. The example I gave is just an example -- in principle, *any* operation can trigger the issue.

**Comment by tamiko** [comment]
> Btw, I don't think it makes a difference what vector type you use. The
> example I gave is just an example -- in principle, *any* operation can
> trigger the issue.

Added a bullet point to the list of things I want to fix (#4992).

**Comment by tamiko** [comment]
As predicted, two know failing tests:
```
The following tests FAILED:
	258 - dofs/interpolate_q_system_mixed_01.debug (Failed)
	2538 - bits/step-15.debug (Failed)
```

**Comment by bangerth** [comment]
Does the original step-15 tutorial also fail? Or what is it about that program?

**Comment by tamiko** [comment]
@bangerth The solver has different initial/final residuals due to the interpolation of starting values. I haven't investigated further.

**Comment by tamiko** [comment]
@bangerth `bits/step-15` has commit changes from you from 2006. example/step-15 was rewritten 2012, see 0b83cd379c9b19636ed89722d3fcda1e96b6b5b5. I doubt they have much in common.

**Comment by bangerth** [comment]
Ah, yes, very different programs indeed. The old step-15 was a much more fickle thing. I'd just update the test results.

**Comment by tamiko** [comment]
@bangerth That will save us for the next two weeks. After I have parallelized `VectorTools::interpolate` we'll be absolutely out of luck :-D

**Comment by tamiko** [comment]
*ping*

**Comment by bangerth** [comment]
Oh, I thought you were going to update test outputs. I'm ok with merging the current state if that was your plan.

**Comment by tamiko** [comment]
@bangerth No, I plan to fix both tests later.

**Comment by bangerth** [comment]
OK then!

**Comment by tjhei** [comment]
> No, I plan to fix both tests later.

For the record: I would prefer we only merge PRs that don't break tests on purpose. This causes a ton of extra work (see the new issues referencing this PR). If you don't want to, you can easily move the tests to fail/ temporarily.

**Comment by tamiko** [comment]
> If you don't want to fix them, you can easily move the tests to fail/ temporarily.

As you wish.

**Comment by bangerth** [comment]
Ugh, another failure caused by this PR: https://github.com/geodynamics/aspect/issues/1936

**Comment by gassmoeller** [comment]
The error messages we see are posted in geodynamics/aspect#1936. In particular it seems linked to MPI-parallel models, and we see them both in FESystems with Q2DGP1, and in models with FESystems with Q2Q1 (we only interpolate on the Q2 component by setting a component mask).
